### PR TITLE
[breaking] - Implement multiple rules and refactor getAttributeValue

### LINKED ISSUE #12: Rule proposal: AX_ARIA_03
I think that this would make a really good linter rule. https://github.com/GoogleChrome/accessibility-developer-tools/wiki/Audit-Rules#ax_aria_03

Elements with ARIA roles must have all required attributes for that role

Good:

```html
<!-- Good: the checkbox role requires the aria-checked state -->
<span role="checkbox" aria-checked="false" aria-labelledby="foo" tabindex="0"></span>
```

Bad:

```html
<!-- Bad: the checkbox role requires the aria-checked state -->
<span role="checkbox" aria-labelledby="foo" tabindex="0"></span>
```

The [WAI-ARIA spec](http://www.w3.org/TR/wai-aria/roles#requiredState) has more information on "Required States and Properties"

### LINKED ISSUE #13: Rule proposal: AX_ARIA_04
I think that this would make a really good linter rule. https://github.com/GoogleChrome/accessibility-developer-tools/wiki/Audit-Rules#ax_aria_04

ARIA state and property values must be valid

The values of ARIA states and properties must be of the correct type.

Bad:

```html
<!-- Bad: the aria-hidden state is of type true/false -->
<span aria-hidden="yes">foo</span>
```

Good:

```html
<!-- Good: the aria-hidden state is of type true/false -->
<span aria-hidden="true">foo</span>
```

### LINKED ISSUE #14: Rule proposal: AX_ARIA_12
I think that this would make a really good linter rule. https://github.com/GoogleChrome/accessibility-developer-tools/wiki/Audit-Rules#ax_aria_12

This element does not support ARIA roles, states and properties

Some HTML elements should not be given ARIA attributes. This is often because they are not visible, for example meta, html, script, style.

Bad:

```html
<!-- Bad: the meta element should not be given any ARIA attributes -->
<meta charset="UTF-8" aria-hidden="false">
```

Good:

```html
<!-- Good: the meta element should not be given any ARIA attributes -->
<meta charset="UTF-8">
```

### PR DESCRIPTION
**Pull request by evcohen**
@lencioni I know this is a huge PR, but all tests are passing and I want to get this code into master so further PR's can be more explicit and modular.

I wanted to refactor `getAttributeValue` so that the extractor functions for different AST attribute value types can be composed together for different use cases (i.e. `getAttributeLiteralValue` will not try and infer values from types that are variables, functions, and other expressions). Those `switch` statements were ugly and harder to debug.

Fixes #12 
Fixes #13 
Fixes #14 
Lays groundwork for #17

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -10,7 +10,8 @@ module.exports = {
     'no-access-key': require('./rules/no-access-key'),
     'label-uses-for': require('./rules/label-uses-for'),
     'no-hash-href': require('./rules/no-hash-href'),
-    'valid-aria-role': require('./rules/valid-aria-role')
+    'valid-aria-role': require('./rules/valid-aria-role'),
+    'no-invalid-aria': require('./rules/no-invalid-aria')
```
Comment (lencioni) [inline src/index.js]: FWIW, this diff would be a little cleaner if you used a trailing comma ;)

File: src/rules/no-invalid-aria.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ */
+'use strict';
+
+// ----------------------------------------------------------------------------
+// Rule Definition
+// ----------------------------------------------------------------------------
+
+import validAriaProperties from '../util/validAriaProperties';
+
+const errorMessage = name => `${name}: This attribute is an invalid ARIA attribute.`;
```
Comment (lencioni) [inline src/rules/no-invalid-aria.js]: As a possible future enhancement, it would be really sweet to suggest attributes that are similar to what they had. I imagine this would really help with typos.

File: src/util/validAriaProperties.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
+// Taken from https://www.w3.org/TR/wai-aria/states_and_properties#state_prop_def
```
Comment (lencioni) [inline src/util/validAriaProperties.js]: Thanks for including comments like this. Links to more information are usually pretty helpful.

File: src/util/validAriaProperties.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  'ARIA-LABELLEDBY', 'ARIA-LEVEL',
+  'ARIA-LIVE', 'ARIA-MULTILINE',
+  'ARIA-MULTISELECTABLE', 'ARIA-ORIENTATION',
+  'ARIA-OWNS', 'ARIA-POSINSET',
+  'ARIA-PRESSED', 'ARIA-READONLY',
+  'ARIA-RELEVANT', 'ARIA-REQUIRED',
+  'ARIA-SELECTED', 'ARIA-SETSIZE',
+  'ARIA-SORT', 'ARIA-VALUEMAX',
+  'ARIA-VALUEMIN', 'ARIA-VALUENOW',
+  'ARIA-VALUETEXT'
```
Comment (lencioni) [inline src/util/validAriaProperties.js]: nit: I would put each value on its own line. That would make finding something you are looking for a bit easier, and make future modifications to this list a bit cleaner.

Comment (evcohen) [inline src/util/validAriaProperties.js]: This gets refactored into object on a later commit. Sorry this PR is a mammoth and hard to follow 🙈 Feel free to reach out if you have any question and checkout the branch and take a look at the code minus the diff.

File: tests/src/rules/no-invalid-aria.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+  parserOptions
+}));
+
+ruleTester.run('no-invalid-aria', rule, {
+  valid: [
+    // Variables should pass, as we are only testing literals.
+    { code: '<div />', parserOptions },
+    { code: '<div></div>', parserOptions },
+    { code: '<div aria="wee"></div>', parserOptions }, // Needs aria-*
+    { code: '<div abcARIAdef="true"></div>', parserOptions },
```
Comment (lencioni) [inline tests/src/rules/no-invalid-aria.js]: Should probably include one that has "aria-" but doesn't start with "aria-", like `fooaria-foo`.

File: src/util/values/expressions/TemplateLiteral.js [hunk 6]
```diff
@@ -1,5 +1,30 @@
-import buildTemplateLiteral from '../../buildTemplateLiteral';
+/**
+ * Returns the string value of a template literal object.
+ * Tries to build it as best as it can based on the passed
+ * prop. For instance `This is a ${prop}` will return 'This is a {prop}'.
+ *
+ * If the template literal builds to undefined (`${undefined}`), then
+ * this should return "".
+ */
+const extract = value => {
```
Comment (lencioni) [inline src/util/values/expressions/TemplateLiteral.js]: You don't really get much out of the arrow functions here. I think this would be better as `export default function extract(value) {`

Comment (lencioni) [inline src/util/values/expressions/TemplateLiteral.js]: Also, ideally these function names match their filenames for maximum readability. I suggest also renaming the filename to be something more like `extractTemplateLiteral`.

File: src/util/values/expressions/UnaryExpression.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
+import getValue from './index';
+
+const extract = value => {
+  const { operator, argument } = value;
+
+  switch (operator) {
+    case '-':
+      return -getValue(argument);
+    case '+':
+      return Number(getValue(argument));
```
Comment (lencioni) [inline src/util/values/expressions/UnaryExpression.js]: Any reason why you chose `Number(getValue(argument))` over `+getValue(argument)`?

File: src/util/values/Literal.js [hunk 8]
```diff
@@ -1,3 +1,13 @@
-const extract = value => value.value;
+const extract = value => {
+  const { value: extractedValue } = value;
+
+  if (extractedValue === "true") {
```
Comment (lencioni) [inline src/util/values/Literal.js]: nit: single quotes

File: src/util/values/index.js [hunk 9]
```diff
@@ -1,15 +1,21 @@
...
   JSXExpressionContainer
 };
 
-const getValue = value => TYPES[value.type](value);
+const LITERAL_TYPES = Object.assign({}, TYPES, {
+  JSXElement: () => null,
+  JSXExpressionContainer: extractLiteral
+});
+
+export const getValue = value => TYPES[value.type](value);
```
Comment (lencioni) [inline src/util/values/index.js]: Similar to my comment on a different commit, I think these would be better off without the arrow functions. In this case `export default function getValue(value) {`

File: src/rules/valid-aria-proptypes.js [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+// ----------------------------------------------------------------------------
+// Rule Definition
+// ----------------------------------------------------------------------------
+
+import ariaAttributes from '../util/ariaAttributes';
+import { getLiteralAttributeValue } from '../util/getAttributeValue';
+
+const errorMessage = (name, type) => `${name} must be of type ${type}.`;
+
+const validityCheck = (value, expectedType, tokens) => {
```
Comment (lencioni) [inline src/rules/valid-aria-proptypes.js]: Since this returns a boolean, it would be nice if it were named something like `isValid()`.

File: src/util/values/Literal.js [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
```
Comment (lencioni) [inline src/util/values/Literal.js]: nit: singles

File: src/util/values/Literal.js [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+    return true;
+  } else if (extractedValue === "false") {
```
Comment (lencioni) [inline src/util/values/Literal.js]: I don't have a lot of context here--does this distinguish between `true`/`false` and `'true'`/`'false'`?

File: tests/src/rules/valid-aria-proptypes.js [hunk 13]
```diff
@@ -25,15 +25,33 @@ const parserOptions = {
...
+    case 'tokenlist':
+      return `The value for ${name} must be a list of one or more tokens from the following: ${permittedValues}.`;
+    case 'boolean':
+    case 'string':
+    case 'integer':
+    case 'number':
+    default:
+      return `The value for ${name} must be a ${type}.`;
+  }
+};
```
Comment (lencioni) [inline tests/src/rules/valid-aria-proptypes.js]: Maybe the rule should export this instead of repeating it in the tests?

File: src/util/values/Literal.js (hunk 12 above)
Comment (evcohen) [inline src/util/values/Literal.js]: Yes for the use case `<div aria-hidden="true" />` we actually want that to resolve to a boolean.

File: src/rules/valid-aria-role.js [hunk 14]
```diff
@@ -9,50 +9,36 @@
...
+    // If value is undefined, then the role attribute will be dropped in the DOM.
+    // If value is null, then getLiteralAttributeValue is telling us that the value isn't in the form of a literal.
+    if (value === undefined || value === null) {
       return;
     }
 
-    // If value is a literal.
-    const normalizedValues = attribute.value.value.toUpperCase().split(" ");
-    const isValid = normalizedValues.every(value => validRoleTypes.indexOf(value) > -1);
+    const normalizedValues = `${value}`.toUpperCase().split(" ");
```
Comment (lencioni) [inline src/rules/valid-aria-role.js]: nit: single quotes

Also, what do you think about using `String(value)` instead of the template literal?

File: src/util/values/expressions/index.js [hunk 15]
```diff
@@ -24,6 +24,13 @@ const TYPES = {
 const noop = () => null;
 
 const LITERAL_TYPES = assign({}, TYPES, {
+  Literal: value => {
+    const extractedVal = TYPES.Literal(value);
```
Comment (lencioni) [inline src/util/values/expressions/index.js]: I think I find this a little opaque. Perhaps it would be better if `TYPES.Literal` was a verb?

File: tests/src/rules/valid-aria-proptypes.js (hunk 13 above)
Comment (evcohen) [inline tests/src/rules/valid-aria-proptypes.js]: I was thinking about extracting all error messages into a single file that just exports an object where key is some unique error code/identifier and value is function/string. This way we only have to change in one place.

File: src/rules/role-requires-aria.js [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+// Rule Definition
+// ----------------------------------------------------------------------------
+
+import validRoleTypes from '../util/attributes/role';
+import { getLiteralAttributeValue } from '../util/getAttributeValue';
+import hasAttribute from '../util/hasAttribute';
+
+const errorMessage = (role, requiredProps) =>
+  `Elements with the ARIA role "${role}" must have the following ` +
+  `attributes defined: ${requiredProps.toString().toLowerCase()}`;
```
Comment (lencioni) [inline src/rules/role-requires-aria.js]: `String(requiredProps)` is preferred to `requiredProps.toString()` in case `requiredProps` gets a `toString` property on it for some reason.

File: tests/src/rules/role-requires-aria.js [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+    { code: '<div />', parserOptions },
+    { code: '<div></div>', parserOptions },
+    { code: '<div role={role} />', parserOptions },
+    { code: '<div role={role || "button"} />', parserOptions },
+    { code: '<div role={role || "foobar"} />', parserOptions },
+    { code: '<div role="tabpanel row" />', parserOptions },
+    { code: '<span role="checkbox" aria-checked="false" aria-labelledby="foo" tabindex="0"></span>', parserOptions },
+    { code: '<Bar baz />', parserOptions }
+  ].concat(basicValidityTests),
+  invalid: [
```
Comment (lencioni) [inline tests/src/rules/role-requires-aria.js]: Scanning this file, it is a bit hard to spot where valid switches over to invalid. It might be nice to add an empty line above this line.

File: src/util/values/expressions/UnaryExpression.js (hunk 7 above)
Comment (evcohen) [inline src/util/values/expressions/UnaryExpression.js]: Had a `no-implicit-coercion` lint error with `+getValue(argument)`

File: src/rules/role-requires-aria.js (hunk 16 above)
Comment (lencioni) [inline src/rules/role-requires-aria.js]: I think this error message would be much better if it just told you the ones you are missing instead of listing all of the ones that are required.

File: src/util/values/Literal.js (hunk 11 above)
Comment (evcohen) [inline src/util/values/Literal.js]: Just for learning sake - why do you prefer singles over doubles?

File: docs/rules/no-unsupported-elements-use-aria.md [hunk 18]
```diff
@@ -0,0 +1,23 @@
+# no-unsupported-elements-use-aria
+
+Certain reserved DOM elements do not support ARIA roles, states and properties. This is often because they are not visible, for example `meta`, `html`, `script`, `style`. This rule enforces that these DOM elements do not contain the role and/or aria-* props.
```
Comment (lencioni) [inline docs/rules/no-unsupported-elements-use-aria.md]: I think you could use backticks around "role" and "aria-*".

File: docs/rules/no-unsupported-elements-use-aria.md [hunk 19]
```diff
@@ -0,0 +1,23 @@
...
+1. [AX_ARIA_12](https://github.com/GoogleChrome/accessibility-developer-tools/wiki/Audit-Rules#ax_aria_12)
+
+## Rule details
+
+This rule takes no arguments.
+
+### Succeed
+```jsx
+<!-- Good: the meta element should not be given any ARIA attributes -->
+<meta charset="UTF-8">
```
Comment (lencioni) [inline docs/rules/no-unsupported-elements-use-aria.md]: This should probably be a self-closing tag because JSX

File: src/rules/no-unsupported-elements-use-aria.js [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+    // If it's not reserved, then it can have ARIA-* roles, states, and properties
+    if (isReservedNodeType === false) {
+      return;
+    }
+
+    // Check if it has role attribute;
+    const hasRole = hasAttribute(node.attributes, 'role');
+    const hasAria = Object.keys(ARIA).some(prop => hasAttribute(node.attributes, prop.toLowerCase()));
+
+    if (hasRole || hasAria) {
```
Comment (lencioni) [inline src/rules/no-unsupported-elements-use-aria.js]: Not sure if it matters, but this would be more efficient if you reported and returned after checking for role, and then checked for aria attributes.

File: src/rules/no-unsupported-elements-use-aria.js [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+    const isReservedNodeType = nodeAttrs && nodeAttrs.reserved || false;
+
+    // If it's not reserved, then it can have ARIA-* roles, states, and properties
+    if (isReservedNodeType === false) {
+      return;
+    }
+
+    // Check if it has role attribute;
+    const hasRole = hasAttribute(node.attributes, 'role');
+    const hasAria = Object.keys(ARIA).some(prop => hasAttribute(node.attributes, prop.toLowerCase()));
```
Comment (lencioni) [inline src/rules/no-unsupported-elements-use-aria.js]: I imagine that there were usually be fewer props than possible aria attributes, so it might be better to invert this loop.

File: tests/src/rules/valid-aria-proptypes.js (hunk 13 above)
Comment (lencioni) [inline tests/src/rules/valid-aria-proptypes.js]: I think it is still valuable for the error messages to be in the rules. It would be useful to put them all in one file if you find yourself working on multiple error messages at the same time, but that seems less likely than iterating on individual rules.

**Comment by lencioni** [comment]
LGTM Thanks for all your work on this!

File: src/util/values/Literal.js (hunk 11 above)
Comment (lencioni) [inline src/util/values/Literal.js]: I mostly prefer consistency, and I've noticed more singles than doubles here.

In general, I like single quotes because in many languages single quotes don't have interpolation, whereas double quotes do. Although there can be a micro micro micro performance benefit in those languages, that's not really important. This gives you a stronger signal of developer intention, which I think makes code more readable. This isn't the case in JS, but it is nice to be consistent for folks who work in multiple languages. And, in any case, they require fewer keypresses.

File: src/util/values/expressions/UnaryExpression.js (hunk 7 above)
Comment (lencioni) [inline src/util/values/expressions/UnaryExpression.js]: Ah interesting. I think it might be worth disabling that rule in this particular case, since you really want to emulate the original code which used the implicit coercion as closely as possible.

Comment (lencioni) [inline src/util/values/expressions/UnaryExpression.js]: BTW, you can disable it for just this line by appending `// eslint-disable-line no-implicit-coercion`

File: tests/src/rules/valid-aria-proptypes.js (hunk 13 above)
Comment (evcohen) [inline tests/src/rules/valid-aria-proptypes.js]: That makes sense. Agreed. Will just export the message/message functions for easier access in tests. We just want to make sure it's reporting the _right_ error message given the test case.

File: src/rules/valid-aria-role.js (hunk 14 above)
Comment (evcohen) [inline src/rules/valid-aria-role.js]: Makes sense. Intention is clearer

File: .eslintrc.js [hunk 22]
```diff
@@ -143,6 +143,7 @@ module.exports = {
     "prefer-const": 2,
     "prefer-spread": 2,
     "prefer-template": 2,
+    'quotes': [2, 'single', 'avoid-escape'],
```
Comment (lencioni) [inline .eslintrc.js]: :+1: but it's really funny to me to see this line amongst a sea of double quotes. :)

File: src/rules/no-unsupported-elements-use-aria.js [hunk 23]
```diff
@@ -26,11 +26,10 @@ module.exports = context => ({
       return;
     }
 
-    // Check if it has role attribute;
-    const hasRole = hasAttribute(node.attributes, 'role');
-    const hasAria = Object.keys(ARIA).some(prop => hasAttribute(node.attributes, prop.toLowerCase()));
+    const invalidAttributes = Object.keys(ARIA).concat('ROLE');
+    const hasAria = hasAttribute(node.attributes, ...invalidAttributes);
```
Comment (lencioni) [inline src/rules/no-unsupported-elements-use-aria.js]: I forget--does hasAttribute do the case conversion? It must

File: .eslintrc.js (hunk 22 above)
Comment (evcohen) [inline .eslintrc.js]: haha just added it for your sanity 😄 
crushing through it right now

File: src/util/hasAttribute.js [hunk 24]
```diff
@@ -7,17 +7,20 @@
  * as the purpose of this linter is to do hard checks of explicit JSX props.
  *
  */
-const hasAttribute = (attributes, attribute) => {
+const hasAttribute = (attributesOnNode, ...attributesToCheck) => {
```
Comment (lencioni) [inline src/util/hasAttribute.js]: Do you know what version of Node added `...`?

File: .eslintrc.js (hunk 22 above)
Comment (lencioni) [inline .eslintrc.js]: Thankfully, eslint can autofix this for you. Just run `eslint . --fix` (or something like that)

File: src/util/hasAttribute.js (hunk 24 above)
Comment (evcohen) [inline src/util/hasAttribute.js]: We transpile as build step, should be good here - the travis config runs on all versions and it just passed

Comment (lencioni) [inline src/util/hasAttribute.js]: Ah, that's right. :+1:

File: package.json [hunk 25]
```diff
@@ -20,7 +20,7 @@
     "build": "rimraf lib && babel src --out-dir lib",
     "prepublish": "npm run lint && npm run test && npm run build",
     "coveralls": "cat ./reports/coverage/lcov.info | coveralls",
-    "lint": "eslint  --config .eslintrc.js src tests",
+    "lint": "eslint  --config .eslintrc.js src/**/*.js tests/**/*.js",
```
Comment (lencioni) [inline package.json]: Instead of globbing like this, I think it would make more sense to use a .eslintignore, and then this becomes `eslint --config .eslintrc.js .`

File: src/util/getAttributeValue.js [hunk 26]
```diff
@@ -1,8 +1,13 @@
 'use strict';
 
-import { getValue, getLiteralValue } from './values';
+import {
+  default as getValue,
```
Comment (lencioni) [inline src/util/getAttributeValue.js]: If you like, I think this can be written as:

```js
import getValue, { getLiteralValue } from './values';
```

Comment (evcohen) [inline src/util/getAttributeValue.js]: ah true, i was struggling with the import there. let me fix. I think good to merge after that

File: src/util/values/index.js [hunk 27]
```diff
@@ -1,22 +1,54 @@
...
+ * @param value - AST Value object on a JSX Attribute.
+ */
+const getLiteralValue = function getLiteralValue(value) {
+  return LITERAL_TYPES[value.type](value);
+};
+
+export {
+  getValue as default,
+  getLiteralValue
+};
```
Comment (lencioni) [inline src/util/values/index.js]: I think you would be better off moving the exports to the function definitions and dropping the variable assignments. Even if you want to keep this here, you could drop the variable assignments since they are pretty redundant in this case.

```js
export function getValue(value) {
...
export default function getLiteralValue(value) {
```

Comment (evcohen) [inline src/util/values/index.js]: Right - this was my workaround to get the import right. Was struggling how to import in this case because getValue was being imported as undefined with this: 

```js
import { getValue, getLiteralValue } from './values';
```
But seeing your comment above about the import clarifies this
